src/Features/LanguageServer/Protocol/Handler/Diagnostics/DocumentPullDiagnosticHandler.cs

```diff
@@ -39,10 +53,18 @@ protected override ImmutableArray<Document> GetOrderedDocuments(RequestContext c
             return context.Document == null ? ImmutableArray<Document>.Empty : ImmutableArray.Create(context.Document);
         }
 
-        protected override DiagnosticParams[]? GetPreviousResults(DocumentDiagnosticsParams diagnosticsParams)
-            => new[] { diagnosticsParams };
+        protected override Task<ImmutableArray<DiagnosticData>> GetDiagnosticsAsync(
+            Document document, Option2<DiagnosticMode> diagnosticMode, CancellationToken cancellationToken)
+        {
+            // We only support doc diagnostics for open files.
+            if (!document.IsOpen())
```

Comment:
It looks like this is using the workspace notion of a document being opened, which is currently tied to the textbuffer being opened via VS.  It is currently true that document textbuffers are opened on the server via VS, but may not be forever (and may not be exactly sync'd with LSP open events similarly to didChange).  

Perhaps instead for the LSP handlers, we should check if we've received a didOpen event for the document (basically if it's a member of this - https://github.com/dotnet/roslyn/blob/master/src/Features/LanguageServer/Protocol/Handler/RequestExecutionQueue.DocumentChangeTracker.cs#L24 ).  Same in the workspace handler

Reply:
yup. i've changed this now.